### DISCUSSION TITLE
Allow formats and style_formats to remove conflicting formats/classes

### DISCUSSION THREAD

**Discussion by colethorsen** [tinymce/tinymce#7581] [Ideas]
I have a regular use case where we use our own classes, rather than setting explicit styles for various different use cases, this allows us to globally adjust the class styles at any point and not have to worry about any specific styles applied via formats in legacy content.

This however leads to an irritant when adding conflicting formats such as text alignment to an selector.

```
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  formats: {
    alignleft: { selector: 'p', classes: 'left' },
    aligncenter: { selector: 'p', classes: 'center' },
    alignright: { selector: 'p', classes: 'right' },
    alignjustify: { selector: 'p', classes: 'full' },
  }
});
```

In the default tinymce way of handling this the `text-align: left;` would be replaced by `text-align: right` if you were to apply right alignment after left alignment, however with classes they just stack so you could end up with `class="left right"`, depending on the order of your styles in your stylesheet it basically makes an arbitrary looking choice. So, in order to change from `left` to `right` you actually have to click off `left` and then click on `right`. This can compound further into a mess like `class="left center right full"`

What if there was an ability to set either `remove_classes` as part of a format or a `remove` object that potentially contains `classes`, `styles`, or event `formats`. I'm not immediately sure on what the use case for explicitly being able to remove styles would be but it would provide more flexibility.

For example:

```
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  formats: {
    alignleft: { selector: 'p', classes: 'left', remove_classes: 'center right full' },
    aligncenter: { selector: 'p', classes: 'center', remove_classes: 'left right full' },
    alignright: { selector: 'p', classes: 'right', remove_classes: 'left center full' },
    alignjustify: { selector: 'p', classes: 'full', remove_classes: 'left center right' },
  }
});
```

or

```
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  formats: {
    alignleft: { selector: 'p', classes: 'left', remove: { classes: 'center right full' } },
    aligncenter: { selector: 'p', classes: 'center', remove: { classes: 'left right full' } },
    alignright: { selector: 'p', classes: 'right', remove: { classes: 'left center full' } },
    alignjustify: { selector: 'p', classes: 'full', remove: { classes: 'left center right' } },
  }
});
```

**Comment by prestonholt** [comment]
@colethorsen Did you happen to find a workaround for this issue? I'm dealing with the same thing

**Comment by colethorsen** [comment]
Hey, I didn't but I haven't really looked into it beyond this post. I'm sure it could be created as a plugin.

**Comment by spocke** [comment]
Could you setup a fiddle where this is happening since I have troubles reproducing this issue. The alignment buttons first remove all the other alignments before applying a new one so you would never end up with two alignment classes at the same time. Use the alignment buttons here and notice that the class of the previous alignment is removed https://fiddle.tiny.cloud/Fkiaab.

If you are using the formatting apis directly I suggest instead using the commands to align things the commands are: `JustifyLeft`, `JustifyCenter`, `JustifyRight`, `JustifyFull`. These commands are where the code exists for first removing the exiting alignments and that is what the alignment buttons use in the UI.

But having a `remove_classes` and `remove_styles` would be useful to avoid having to first remove all the existing alignments before applying a new one. But so far alignment has been the only case where this has been required.

**Comment by colethorsen** [comment]
@spocke use this fiddle and try using the style select dropdown instead https://fiddle.tiny.cloud/Fkiaab/1

Also there are many cases where this could be useful beyond justification. For instance we use a number of sizing styles to allow for the scale up or scale down of text, they're roughly named `.text_sm`, `.text_md`, `.text_lg`, `.text_xl`, currently, if text is styled to a size, a user must un-select that size and then re-select the next one, we also have color options that follow a similar pattern, where a user can choose classes `.color_primary`, `.color_secondary`, `.color_tertiary`, again a user must un-select an existing color before choosing a new one. 

The TinyMCE "solution" to this is basically to use styles so the color would be set as a hard coded value, or the size would be set in px, em, rem, etc. This isn't really suitable as its much more effective to abstract the styling to classes as they can then be globally controlled, and effected by other factors, such as device, other elements this content may be nested in, etc.

There are also numerous more abstract situations where this would be helpful, like having several different styling options set up for text where they are direct replacements for each other and wouldn't be used in combination, like say pullquote styling, often we provide several different complimentary style options for this, but you'd never want to combine the styles for them, picking a different class just adds it, which can cause all sorts of crazy unintended consequences for users.

A clearer example (give that the formatting does seem to work in your example, just not in mine when implemented in a different way) would probably be:

```
tinymce.init({
    selector: 'textarea',  // change this value according to your html
    style_formats: { title : "Sizing", items : [
        { title : "XL", selector : "p", classes : "text_xl", remove : { classes : "text_lg text_md text_sm" } },
        { title : "Large", selector : "p", classes : "text_lg", remove : { classes : "text_xl text_md text_sm" } },
        { title : "Medium", selector : "p", classes : "text_md", remove : { classes : "text_xl text_lg text_sm" } },
        { title : "Small", selector : "p", classes : "text_sm", remove : { classes : "text_xl text_lg text_md" } },
    ] },
});
```

**Reply by tvdeyen** [reply]
@spocke we have the exact same issue. 

It would even be working for us if the `selector` option really would allow to use CSS3 selectors ([as claimed in the documentation](https://www.tiny.cloud/docs/tinymce/latest/content-formatting/#selector)),
but it does not allow CSS3 selectors (just tag selectors. It raises an error  in `createElement` if you try to use CSS3 selectors). 
Even if combined with `block` it does not work as expected (and documented). A real bummer (and probably a bug).

Anyhow, if this bug would be fixed, we could at least do:

```js
{ block: "p", selector: "p:not(.text_sm)", classes: "text_xl" },
{ block: "p", selector: "p:not(.text_xl)", classes: "text_sm" }
```

Also this would be very verbose and error prone. So not ideal.

A much better idea imo would be to be able to set something like `exclusive: true`:

```js
{ selector: "p", classes: "text_xl", exclusive :true },
{ selector: "p", classes: "text_sm", exclusive :true }
```

Or even better in `style_formats`

```js
style_formats: {
  title: "Text Size",
  exclusive: true, // or inclusive: false
  items: [
    { selector: "p", classes: "text_xl" },
    { selector: "p", classes: "text_sm" }
  ]
}, {
  title: "Text Additions",
  exclusive: false, // or inclusive: true
  items: [
    { selector: "p", classes: "text_bold" },
    { selector: "p", classes: "text_italic" }
  ]
}
```

Then users would be able to have two groups of formats. Formats that just adds classes and formats that toggle classes.

I am very certain that other users having the same problem, especially with TailwindCSS.

**Comment by tvdeyen** [comment]
OMG, just found this https://stackoverflow.com/a/56224417/527689

> p_grey: { selector: 'p', attributes: {'class':'grey'} }, // use attributes

Using `attributes` replaces the `class` attribute. This is S.M.A.R.T.

But still it would be great to not have to do a workaround like this.

**Comment by colethorsen** [comment]
Thats a good solution. I ended up writing a solution for this too which basically solves the problem exactly as initially described, on the surface the above solution looks like it relies on the nesting of menus, this one just uses the prefix of the formats regardless of where they are applied from. It takes the approach of defaulting to toggling, and then utilizing a `custom-` prefixed format to disable the toggling. It also works on headers (I decouple styling from tag for headers in order to to promote proper use of header order instead of choosing headers based on how they look)

```javascript
const config = {
	formats: {
		align_left   : { selector : 'p,h1,h2,h3,h4,h5,h6,td,th,div,ul,ol,li,table', classes : '' },
		align_center : { selector : 'p,h1,h2,h3,h4,h5,h6,td,th,div,ul,ol,li,table', classes : 'text_center' },
		align_right  : { selector : 'p,h1,h2,h3,h4,h5,h6,td,th,div,ul,ol,li,table', classes : 'text_right' },

		h1 : { block : 'h1', classes : 'bold' },
		h2 : { block : 'h2', classes : 'bold' },
		h3 : { block : 'h3', classes : 'bold' },
		h4 : { block : 'h4', classes : 'bold' },
		h5 : { block : 'h5', classes : 'bold' },
		h6 : { block : 'h6', classes : 'bold' },

		header_sm : { selector : 'h1,h2,h3,h4,h5,h6', classes : 'header_sm' },
		header_md : { selector : 'h1,h2,h3,h4,h5,h6', classes : 'header_md' },
		header_lg : { selector : 'h1,h2,h3,h4,h5,h6', classes : 'header_lg' },
		header_xl : { selector : 'h1,h2,h3,h4,h5,h6', classes : 'header_xl' },

		color_primary   : { inline : 'span', classes : 'text_primary' },
		color_secondary : { inline : 'span', classes : 'text_secondary' },
	},
	style_formats : [
		{ title : "Headers", items : [
			{ title : "Header 1", format : "h1" },
			{ title : "Header 2", format : "h2" },
			{ title : "Header 3", format : "h3" },
			{ title : "Header 4", format : "h4" },
			{ title : "Header 5", format : "h5" },
			{ title : "Header - XL", format : "header_xl" },
			{ title : "Header - Large", format : "header_lg" },
			{ title : "Header - Medium", format : "header_md" },
			{ title : "Header - Small", format : "header_sm" }
		] },
		{ title : "Inline", items : [
			{ title : 'Left', icon : 'align-left', format : 'align_left' },
			{ title : 'Center', icon : 'align-center', format : 'align_center' },
			{ title : 'Right', icon : 'align-right', format : 'align_right' },
			{ title : 'Uppercase', inline : 'span', classes : 'uppercase' },
		] },
		{ title : "Colours", items : [
			{ title : 'Primary', format : 'color_primary' },
			{ title : 'Secondary', format : 'color_secondary' },
		] },
	],
	setup : function (ed) {
		//allows us to have "toggleable" formats for instance with the header sizes and aligns.
		ed.on('ExecCommand', function checkListNodes (evt) {

			if (evt.command == 'mceToggleFormat') {

				//don't do anything for custom.
				if (evt.value.startsWith('custom-'))
					return true;

				//detect if the current format is already applied or not. so we can know
				//whether or not to re-apply it at the end.
				//
				//We strip off all of the formats and then re-add the selected format if it
				//wasn't being toggled off this way if formats share CSS classes they won't
				//get accidentally removed when we remove the other matching formats.
				const formatAdded = this.formatter.match(evt.value);

				//applying this format before adding/removing others ensures that the
				//element is not rolled up into its parent.
				this.formatter.apply('toggle_fix');

				//deal with header formats as separately as they don't have an underscore in
				//them but still need to be removed in case they have classes associated
				//with them
				const headers = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'];

				if (headers.includes(evt.value)) {

					headers.forEach((format) => {
						if (format != evt.value) {

							//since we've already changed the tag, the format is considered
							//removed even though the classes weren't removed. So we have
							//to re-apply them before removing them.
							this.formatter.apply(format);
							this.formatter.remove(format);
						}
					});

				} else {
					const currentFormat = evt.value.split('_');

					if (currentFormat.length < 2)
						return true;

					//iterate over all the formats.
					Object.entries(wysiwyg_formats).forEach(([format, formatInfo]) => {

						//identify the formats that have the same prefix as the format that
						//was just set, but aren't the one that was just set.
						if (format.startsWith(currentFormat[0] + '_') && format != evt.value) {

							//remove formats that have the same prefix.
							this.formatter.remove(format);
						}
					});
				}

				if (formatAdded)
					this.formatter.apply(evt.value);

				//remove the toggle fix as its no longer necessary.
				this.formatter.remove('toggle_fix');
			}
		});
	},
}
```
